flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/delegation/PlannerContext.java

```diff
@@ -80,13 +82,15 @@
  */
 @Internal
 public class PlannerContext {
+
 	private final RelDataTypeSystem typeSystem = new FlinkTypeSystem();
 	private final FlinkTypeFactory typeFactory = new FlinkTypeFactory(typeSystem);
 	private final TableConfig tableConfig;
-	private final RelOptCluster cluster;
 	private final FlinkContext context;
 	private final CalciteSchema rootSchema;
 	private final List<RelTraitDef> traitDefs;
+	private FrameworkConfig frameworkConfig;
+	private RelOptCluster cluster;
```

Comment:
Can these member fields to be `final`?

Reply:
There is a factory refer these fields before their initialization. According to the semantics of Java, it needs to be non final.

Reply:
We can have a `private SqlExprToRexConverter createSqlExprToRexConverter(RelDataType rowType)` method and pass `this::createSqlExprToRexConverter` to the `FlinkContextImpl`, then these fields can be `final`.

Reply:
I am not favor of it.
Pass `this::createSqlExprToRexConverter` to the `FlinkContextImpl` means we pass an unfinished instance to it. Actually `frameworkConfig` and `cluster` is not the real final to `FlinkContextImpl`, they will become real instance from null.
So this feel like bypassing the `final` rule of Java.

Reply:
The code is a little messy, `FrameworkConfig` reference the `cluster`, while the `cluster` context have a member that needs a `FrameworkConfig` and the `cluster` itself ...

We may need to re-think if the `SqlExprToRexConverter` should belong to the `cluster` context, can we just pass in the `FrameworkConfig` into the context and constructing the `SqlExprToRexConverter` when we really need that (i.e. the `CatalogSourceTable`)?

Reply:
Hi @danny0405 , The constructor of `FrameworkConfig` need cluster context.

Context need a factory which need `FrameworkConfig`. IMO, relationship like this is very common. That's one of the reasons we need the factory model.
I admit we can improve it, but if it's a big refactoring, it's unnecessary, because this PR is just an urgent bug fixing.
Of course, if you have a better and slight improvement plan, you can describe it detailed.

What do you think?

Reply:
After some research of the code, i found another solution, that is:

- Modify the FlinkRelBuilder to add a new method `scan(Iterable<String> tableNames, ViewExpander viewExpander)`, then in the method, fetch the `RelOptTable` first and then do the logic same with `TableScanRule`
- In the `QueryOperationConverter`, use the new `scan` method to translate the table
- No other code change is needed

We do not need to pass the `SqlToExprConverter` around through the `FlinkContext` because the it does not belong there, the computed column/watermark translation should happen as early as possible and not in planner rules.

Reply:
Hi @wuchong , Made PlannerContext fields final.

Reply:
After discussing with @danny0405  offline, We can improve this to finish solution#3 in 1.11.